node/primitives/src/lib.rs

```diff
@@ -56,6 +57,112 @@ impl Statement {
+enum CompactStatement {
+	/// A statement about a new candidate being seconded by a validator. This is an implicit validity vote.
+	#[codec(index = "1")]
+	Seconded(Hash),
+	/// A statement about the validity of a candidate, based on candidate's hash.
+	#[codec(index = "2")]
+	Valid(Hash),
+	/// A statement about the invalidity of a candidate.
+	#[codec(index = "3")]
+	Invalid(Hash),
+}
+
+impl EncodeAs<CompactStatement> for Statement {
+	fn encode_as(&self) -> Vec<u8> {
+		let statement = match *self {
+			Statement::Seconded(ref c) => {
+				polkadot_primitives::parachain::Statement::Candidate(c.hash())
+			}
+			Statement::Valid(hash) => polkadot_primitives::parachain::Statement::Valid(hash),
+			Statement::Invalid(hash) => polkadot_primitives::parachain::Statement::Invalid(hash),
+		};
+		statement.encode()
+	}
+}
+
+/// A signed type which encapsulates the common desire to sign some data and validate a signature.
+///
+/// Note that the internal fields are not public; they are all accessable by immutable getters.
+/// This reduces the chance that they are accidentally mutated, invalidating the signature.
+pub struct Signed<Payload, RealPayload = Payload> {
+	/// The payload is part of the signed data. The rest is the signing context,
+	/// which is known both at signing and at validation.
+	payload: Payload,
+	/// The index of the validator signing this statement.
+	validator_index: ValidatorIndex,
+	/// The signature by the validator of the signed payload.
+	signature: ValidatorSignature,
+	/// This ensures the real payload is tracked at the typesystem level.
+	real_payload: std::marker::PhantomData<RealPayload>,
+}
+
+// We can't bound this on `Payload: Into<RealPayload>` beacuse that conversion consumes
+// the payload, and we don't want that. We can't bound it on `Payload: AsRef<RealPayload>`
+// because there's no blanket impl of `AsRef<T> for T`. In the end, we just invent our
+// own trait which does what we need: EncodeAs.
+impl<Payload: EncodeAs<RealPayload>, RealPayload: Encode> Signed<Payload, RealPayload> {
+	fn payload_data(payload: &Payload, context: SigningContext) -> Vec<u8> {
+		(payload.encode_as(), context).encode()
+	}
+
+	pub fn sign(
+		payload: Payload,
+		context: SigningContext,
+		validator_index: ValidatorIndex,
+		key: &ValidatorPair,
+	) -> Self {
+		let data = Self::payload_data(&payload, context);
+		let signature = key.sign(&data);
+		Self {
+			payload,
+			validator_index,
+			signature,
+			real_payload: std::marker::PhantomData,
+		}
+	}
+
+	pub fn validate(&self, context: SigningContext, key: &ValidatorId) -> bool {
+		let data = Self::payload_data(&self.payload, context);
+		self.signature.verify(data.as_slice(), key)
+	}
+
+	pub fn payload(&self) -> &Payload {
+		&self.payload
+	}
+
+	// REVIEW: this works because ValidatorIndex is Copy, but that's dissimilar to the
+	// other getters. Keep it like this, or borrow it?
+	pub fn validator_index(&self) -> ValidatorIndex {
+		self.validator_index
```

Comment:
i.e. should we change this to

```suggestion
	pub fn validator_index(&self) -> &ValidatorIndex {
		&self.validator_index
```

Reply:
Types smaller or equal to the word size make no sense to return or pass them as reference.